File: 946-validate-stack-sequences/946-validate-stack-sequences.py
from collections import deque
class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        stack = []
        pushed = deque(pushed)
        popped = deque(popped)
        
        while pushed or popped:
            # print(stack,pushed,popped)
            if stack and popped and stack[-1]==popped[0]:
                popped.popleft()
                stack.pop()
            elif pushed:
                stack.append(pushed.popleft())
            else:
                return False
                
        # print(pushed,popped)
        
        return True
        
        

File: 946-validate-stack-sequences/946-validate-stack-sequences.js
/**
 * @param {number[]} pushed
 * @param {number[]} popped
 * @return {boolean}
 */
function validateStackSequences(pushed, popped) {
    let i=0;
    const stack=[];
    for(let ele of pushed){
        stack.push(ele);
        
        while(stack.length && stack.at(-1)===popped[i]){
            stack.pop();
            i++;
        }
    }
    return i===popped.length;
}
